lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -410,36 +404,43 @@ def test_put_invalid_deletion(self, query_get_as, query_put_as):
         }
 
     def test_put_set_errors(self, monkeypatch, query_get_as, query_put_as):
-        real_set = Metadata.setvalue
-
-        @staticmethod
-        def fake_set(dataset: Dataset, key: str, value: Any, user_id: str):
-            if key in ("global.dashboard.seen", "dataset.name"):
-                raise MetadataBadStructure(dataset, key, key)
-            real_set(dataset, key, value, user_id)
-
-        monkeypatch.setattr(Metadata, "setvalue", fake_set)
+        """Test a partial success. We set a scalar value on a key and then try
+        to set a nested value: i.e., with "global.dashboard.nested = False", we
+        attempt to set "global.dashboard.nested.dummy". We expect this to fail,
+        but we expect other values to succeed. This partial success is reported
+        as an "internal error" (for lack of any better representation in HTTP)
+        with a message indicating the specific failures.
```

Comment:
Is the "internal error" generated by the test code or by the Server code?  (It's OK for the test code, but for the Server code it seems like we ought to be able to come up with some sort of alternative which points to a client-side problem rather than suggesting a Server-side problem, at least for non-Server metadata keys....)

Reply:
There are several issues at play here, and I've run into this issue several places in the API (e.g., delete/publish partial completion). There are "internal failures" which might occur here, which might possibly be retryable. That's not something that the HTTP error space handles well, and in previous discussions we've decided that a 500 is the best (or at least "least worst") representation, especially as 500 class errors may be retryable while no relevant 400 class error is.

However when I discovered how easy it is to provoke the sort of error I wanted here, I also realized that we could and possibly should distinguish the deliberately detected `MetadataBadStructure` from truly unexpected internal errors. But I have no idea how to "reasonably" represent that. It could be a 400 class error, because it's only retryable if you change the structure of the existing metadata. I've thought several times of pushing the multi-key set down so that I could abort before anything is changed, or making a pre-validate, but that never really seemed worth the effort given the other things that need to be done.

I've considered strategies such as handling the `MetadataBadStructure` and failing with `CONFLICT` if there's a single key, or maybe even if we find it on the *first* key we set; but it doesn't capture the possibility that other changes have already been made. With some massive and ugly refactoring (and if we didn't have to worry about two radically different DB engines) the whole thing could be built as a nested transaction that could be rolled back.

In any case, if you have a better idea than we've previously invented for "partial success that might be retryable" than 500, that's an alternative we can consider, but I'd rather not get into any of that for *this* PR because it's a lot more complicated than just changing the returned message string.

Reply:
TODO: [PBENCH-1009](https://issues.redhat.com/browse/PBENCH-1009) addresses this.

Reply:
Can you post this as a GitHub issue instead?  As the project lead I am not in favor of having non-public URLs referenced as links in PRs and issues.